runtime/src/main/cpp/Memory.cpp

```diff
@@ -1604,37 +1629,75 @@ void garbageCollect(MemoryState* state, bool force) {
   if (g_hasCyclicCollector)
     cyclicLocalGC();
 #endif  // USE_CYCLIC_GC
+#if PROFILE_GC
+  auto processDecrementsStartTime = konan::getTimeMicros();
+#endif
   processDecrements(state);
+#if PROFILE_GC
+  auto processDecrementsDuration = konan::getTimeMicros() - processDecrementsStartTime;
+  GC_LOG("||| GC: processDecrementsDuration = %lld\n", processDecrementsDuration);
+  auto decrementStackStartTime = konan::getTimeMicros();
+#endif
   size_t beforeDecrements = state->toRelease->size();
   decrementStack(state);
   size_t afterDecrements = state->toRelease->size();
+#if PROFILE_GC
+  auto decrementStackDuration = konan::getTimeMicros() - decrementStackStartTime;
+  GC_LOG("||| GC: decrementStackDuration = %lld\n", decrementStackDuration);
+#endif
   long stackReferences = afterDecrements - beforeDecrements;
   if (state->gcErgonomics && stackReferences * 5 > state->gcThreshold) {
     increaseGcThreshold(state);
-    GC_LOG("||| GC: too many stack references, increased threshold to \n", state->gcThreshold);
+    GC_LOG("||| GC: too many stack references, increased threshold to %d\n", state->gcThreshold);
   }
 
   GC_LOG("||| GC: toFree %d toRelease %d\n", state->toFree->size(), state->toRelease->size())
-
+#if PROFILE_GC
+  auto processFinalizerQueueStartTime = konan::getTimeMicros();
+#endif
   processFinalizerQueue(state);
+#if PROFILE_GC
+  auto processFinalizerQueueDuration = konan::getTimeMicros() - processFinalizerQueueStartTime;
+  GC_LOG("||| GC: processFinalizerQueueDuration %lld\n", processFinalizerQueueDuration);
+#endif
 
-  if (force || state->toFree->size() > kMaxToFreeSize) {
+  if (force || state->toFree->size() > state->gcCollectCyclesThreshold) {
+    auto cyclicGcStartTime = konan::getTimeMicros();
     while (state->toFree->size() > 0) {
       collectCycles(state);
+      #if PROFILE_GC
+        processFinalizerQueueStartTime = konan::getTimeMicros();
+      #endif
       processFinalizerQueue(state);
+      #if PROFILE_GC
+        processFinalizerQueueDuration += konan::getTimeMicros() - processFinalizerQueueStartTime;
+        GC_LOG("||| GC: processFinalizerQueueDuration = %lld\n", processFinalizerQueueDuration);
+      #endif
+    }
+    auto cyclicGcEndTime = konan::getTimeMicros();
+    #if PROFILE_GC
+      GC_LOG("||| GC: collectCyclesDuration = %lld\n", cyclicGcEndTime - cyclicGcStartTime);
+    #endif
+    auto cyclicGcDuration = cyclicGcEndTime - cyclicGcStartTime;
+    if (state->gcErgonomics && cyclicGcDuration > kGcCollectCyclesMinimumDuration &&
+        double(cyclicGcDuration) / (gcStartTime - state->lastCyclicGcTimestamp + 1) > kGcCollectCyclesLoadRatio) {
```

Comment:
`s/gcStartTime/cyclicGcStartTime/`?

Reply:
I amn't sure  because we wanted time of program execution, in this case we get time of program execution + part of GC. But at the same time we don't subtract time of other GC without cyclic collection. Ok, I'll change for consistency.